lib/utils/report_output.rb

```diff
@@ -0,0 +1,63 @@
+require "securerandom"
+
+# An attempt at simplifying the structure of report output dirs and files.
+# Give report_name as input.
+# The methods and their return:
+#   id: the "unique-ish" part of the filename
+#   dir: path to an output dir
+#   file: path to an output file
+#   handle("r"): an open read handle for file
+#   handle("w"): an open write handle for file
+# Both dir and file will contain the report name,
+# and the file a timestamp and a unique-ish string.
+# dir is autovivified with mkdir_p.
+# Example:
+# output = Utils::ReportOutput.new("foo", ".txt")
+# output.dir # -> "/local_reports/foo"
+# output.file # -> "/local_reports/foo/foo_YYYYMMDD_abcd1234.txt"
+
+module Utils
+  class ReportOutput
+    attr_reader :report_name, :ext
+    def initialize(report_name, ext = ".tsv")
+      @report_name = report_name # e.g. "cost_report" or "cost_report_umich"
+      @ext = ext
+    end
+
+    def file
+      if @file.nil?
+        @file = File.join(dir, id) + ext
+      end
+      @file
+    end
+
+    def dir(mkdir: true)
+      # In testing it is useful to get the path without mkdir_p,
+      # hence the mkdir param
+      dir_path = File.join(Settings.local_report_path || "/tmp", report_name)
```

Comment:
Could we do `@dir ||= File.join(...)` here?

Worth considering that right now, if you call this once without `mkdir` set, but then call it again later after changing `Settings.local_report_path` and with `mkdir` set, it will return the existing `@dir` but still make `dir_path`. Not sure that's intended.

Reply:
Actually in looking at the test I would recommend removing the `mkdir` parameter and doing

```ruby
def dir()
  @dir ||= File.join(Settings.local_report_path || "/tmp", report_name).tap do |d|
    FileUtils.mkdir_p(dir_path)
  end
end
```

See separate comment below about the test

Reply:
Yeah. Broke out dir into 2 methods and got rid of mkdir param. Close to what you suggested but not quite.